cdap-app-templates/cdap-program-report/src/main/scala/co/cask/cdap/report/RecordAggregator.scala

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.report
+
+import co.cask.cdap.report.util.Constants
+import org.apache.spark.sql.{Encoders, Row}
+import org.apache.spark.sql.expressions.Aggregator
+
+/**
+  * An aggregator that aggregates [[Row]]'s with the same program run ID into intermediate [[RecordBuilder]]'s
+  * and finally merge the [[RecordBuilder]]'s to build a single [[Record]]
+  */
+class RecordAggregator extends Aggregator[Row, RecordBuilder, Record] {
+
+  def zero: RecordBuilder = RecordBuilder("", "", "", Vector.empty, None)
+  def reduce(builder: RecordBuilder, row: Row): RecordBuilder = {
+    // Get the StartInfo from the builder if it exists or construct a new StartInfo from the row
+    val startInfo: Option[StartInfo] = if (builder.startInfo.isDefined) builder.startInfo else {
```

Comment:
try not to use `if-else` on `Option`. Use `getOrElse`.

Reply:
since here I need an option from option, and I haven't found a clean way to convert the part in `else` from `Option[StartInfo] to a `StartInfo`, so I'll probably keep it as this

Reply:
Use `orElse`.

// Assuming you have the rowToStartInfo private method defined
`val startInfo = builder.startInfo.orElse(Option(row.getAs[Row](Constants.START_INFO)).map(rowToStartInfo(_)))`